openmc/mgxs/mgxs.py

```diff
@@ -5011,13 +5028,1153 @@ class NuFissionMatrixXS(MatrixMGXS):
+        # The tally slicing is needed if tally merging was used
+        for tally_type, tally in self.tallies.items():
+            sp_tally = statepoint.get_tally(
+                tally.scores, tally.filters, tally.nuclides,
+                estimator=tally.estimator, exact_filters=True)
+            sp_tally = sp_tally.get_slice(
+                tally.scores, filters, filter_bins, tally.nuclides)
+            sp_tally.sparse = self.sparse
+            self._tallies[tally_type] = sp_tally
+
+        # Load data for the tallies in each MGXS in the convolution
+        for mgxs in self.mgxs:
+            mgxs.load_from_statepoint(statepoint)
+
+        self._loaded_sp = True
+
+
+class ConsistentScatterMatrixXS(ConvolvedMGXS, ScatterMatrixXS):
+    r"""A scattering matrix multi-group cross section computed as the product
+    of the scatter cross section and group-to-group scattering probabilities.
+
+    This class is a variation of the :class:`ScatterMatrixXS` which computes
+    the scattering matrix as the convolution product of :class:`ScatterXS` and
+    :class:`ScatterProbabilityMatrix`. Unlike the :class:`ScatterMatrixXS`,
+    this scattering matrix is computed from the scattering cross section which
+    uses a tracklength estimator. This ensures that reaction rate balance is
+    exactly preserved with a :class:`TotalXS` computed using a tracklength
+    estimator.
+
+    This class can be used for both OpenMC input generation and tally data
+    post-processing to compute spatially-homogenized and energy-integrated
+    multi-group cross sections for multi-group neutronics calculations. At a
+    minimum, one needs to set the
+    :attr:`ConsistentScatterMatrixXS.energy_groups` and
+    :attr:`ConsistentScatterMatrixXS.domain` properties. Tallies for the flux
+    and appropriate reaction rates over the specified domain are generated
+    automatically via the :attr:`ConsistentScatterMatrixXS.tallies` property,
+    which can then be appended to a :class:`openmc.Tallies` instance.
+
+    For post-processing, the :meth:`MGXS.load_from_statepoint` will pull in the
+    necessary data to compute multi-group cross sections from a
+    :class:`openmc.StatePoint` instance. The derived multi-group cross section
+    can then be obtained from the :attr:`ConsistentScatterMatrixXS.xs_tally`
+    property.
+
+    For a spatial domain :math:`V`, incoming energy group
+    :math:`[E_{g'},E_{g'-1}]`, and outgoing energy group :math:`[E_g,E_{g-1}]`,
+    the Legendre scattering moments are calculated as:
+
+    .. math::
+
+       \langle \sigma_{s,g'\rightarrow g} \phi \rangle &= \int_{r \in V} dr
+       \int_{4\pi} d\Omega' \int_{E_{g'}}^{E_{g'-1}} dE' \int_{4\pi} d\Omega
+       \int_{E_g}^{E_{g-1}} dE \; \sigma_s (r, E'
+       \rightarrow E, \Omega' \cdot \Omega) \psi(r, E', \Omega')\\
+       \langle \phi \rangle &= \int_{r \in V} dr \int_{4\pi} d\Omega
+       \int_{E_g}^{E_{g-1}} dE \; \psi (r, E, \Omega) \\
```

Comment:
I'm talking about the limits of integration on energy. Written as is, you have the g' -> g rate divided by the group g flux instead of the group g' flux.